### DISCUSSION TITLE
Publish changelog to Confluence

### DISCUSSION THREAD

**Discussion by anyulled** [jreleaser/jreleaser#912] [Ideas]
Hi Andres, I'm on a new company and realized that teams here generate a changelog in a markdown file and also publish a confluence page whenever they release a new version. I have already started a branch in my repo to do this integration, as I believe there's no similar way to do it nowadays with jReleaser, am I right?

**Comment by aalmiray** [comment]
Correct, there's no explicit support for uploading files to Confluence, unless you count the `HttpUploader` assuming Confluence allows arbitrary `POST` or `PUT`.

Perhaps adding a `ConfluenceUploader` may be a good idea. This would allow uploading files/artifacts as attachments to a given page. One of the input files should be the one used for updating the page's content (body). This could be done by marking a particular file with an extraProperty such as `confluencePage`. If additional data were to be required per file or attachment then more extra properties may be defined. WDYT?

**Comment by anyulled** [comment]
Thanks for your swift reply, I was thinking of using an announcer because if you don't specify a message it will use the changelog, right?

My idea is to take this changelog and apply an HTML transformation with simple unordered list and include the text in the POST message, this is a typical request to create a new confluence page:

```
POST https://sample.atlassian.net/wiki/rest/api/content
Authorization: Basic user@email.com :privateAccessToken
Accept: application/json
Content-Type: application/json

{
  "type": "page",
  "title": "My Test Page",
  "space": {
    "key": "TEST"
  },
  "body": {
    "storage": {
      "value": "JReleaser is now version 1.2.0. Here's a summary of the changes: ",
      "representation": "storage"
    }
  }
}
```

Meanwhile, I will check the classes you mentioned in your comment to modify my code as you suggested.

**Reply by aalmiray** [reply]
Hmm not really. Announcers use whatever data you explicitly pass them. The changelog is not passed by default, you must do so explicitly. There's a new `{{f_md2html}}` function coming up in `1.2.0` that can transform markdown into HTML (see https://jreleaser.org/guide/early-access/configuration/name-templates.html).

I see now a few options ahead of us:
- ConfluenceAnnouncer: creates a new page on `announce`
- ConfluenceUploader: creates a new page on `upload`. Publishes files/artifacts as attachments. Somehow the body of the page should be specified. Either the uploader defines a explicit property for it (like Teams does with `messageTemplate`) or one of the input files has an extraProperty that marks it as the page's content.

Work on these two may be reused later for setting up Bitbucket (or Atlassian) as a releaser as we could use a Confluence page as release page.

**Reply by aalmiray** [reply]
Filed https://github.com/jreleaser/jreleaser/issues/921 as I think it would be a good idea to get a generic HTTP announcer, similarly as it's done for webhooks.

**Reply by aalmiray** [reply]
Alright, got something done for #921. The bits should be ready with the next `early-access` release.
Have a look at the Confluence example, perhaps this does what you need 😅 

https://jreleaser.org/guide/early-access/configuration/announce/http.html#_confluence

**Comment by anyulled** [comment]
Sweet! I'll give it a try this week and tell you something. Many thanks!
